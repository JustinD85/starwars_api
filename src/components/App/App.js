import React, { Component } from 'react'
import './App.scss'
import ScrollingText from '../ScrollingText/ScrollingText'
import FilterSection from '../FilterSection/FilterSection'
import CardArea from '../CardArea/CardArea'
import API from '../../api/api'
class App extends Component {
  constructor() {
    super()
    this.state = {
      film: {},
      favorites: [],
      currentFilter: null,
      min: 0,
      max: 0,
      error: null
    }
    this.audio = React.createRef()
  }

  setRandomFilm = async () => {
    try {
      const unfilteredData = await API.getData('films')
      const films = unfilteredData.results
      const { length } = films
      const index = Math.floor(Math.random() * length)
      this.setState({ film: { ...films[index] } })
    } catch (error) {
      this.setState({
        error: error.message,
      })
    }
  }

  toggleFavorite = (url) => {
    let favCopy = this.state.favorites.slice()
    favCopy.includes(url) ?
      favCopy.splice(favCopy.indexOf(url), 1)
      :
      favCopy.push(url)
    localStorage.setItem('favorites', JSON.stringify(favCopy))
    this.setState({ favorites: favCopy })
  }

  setCurrentFilter = (type) => {
    this.setState({
      currentFilter: type,
      min: 0,
      max: 8
    })
  }

  changeNumber = (change) => {
    this.setState({
      min: this.state.min + change,
      max: this.state.max + change,
    })
  }

  componentDidMount() {
    this.setState({ favorites: JSON.parse(localStorage.getItem('favorites')) || [] })
    this.setRandomFilm()
  }

  render() {
    const { film, currentFilter, min, max, favorites } = this.state

    return (
      <div className="App">
        <ScrollingText film={film} />
        <FilterSection
          setCurrentFilter={this.setCurrentFilter}
          favoriteCount={favorites.length}
          currentFilter={currentFilter}
        />
        <CardArea
          currentFilter={currentFilter}
          getData={API.getData}
          min={min}
          max={max}
          audio={this.audio}
          favorites={[...favorites]}
          changeNumber={this.changeNumber}
          toggleFavorite={this.toggleFavorite}
        />
        <audio ref={this.audio}><source src="./beep4.mp3"></source></audio>
        <img className="background" src="./sw_background.jpg" alt="shows deathstar in background" />
      </div>
    )
  }
}

export default App
