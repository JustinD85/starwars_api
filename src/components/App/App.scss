
.App {
  color:gold;
  height:100vh;
  width:80vw;
  margin: 0 auto;
  display:flex;
  flex-direction: column;
  align-items: center;

  .filter-buttons{
    height:30%;
    display: flex;
    button{
      width:300px;
    }


   
  }
  .card-area{
    height:50%;
    background:grey;
    width:100%;
    display:flex;
    justify-content: center;
    flex-wrap: wrap;
    div{
      width:30%;
      text-align: center;

    }
  }
}