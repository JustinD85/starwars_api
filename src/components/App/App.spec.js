import React from 'react';
import App from './App';
import { shallow } from 'enzyme';

describe('App', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  it('renders without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  

  it('should fetch films from api', async () => {
    //mock fetch
    expect(wrapper.state('films').length).toEqual(0)
    await wrapper.instance().getFilms();
    expect(wrapper.state('films').length).toEqual(7)
  });

  it('should get random film scroll text,year, and title', () => {
    //Ask best way to test random number generator
  });

  // it('should have 3 props of correct type', async () => {

  //   await wrapper.instance().getFilms();
  //   expect(wrapper.find(ScrollingText).prop('film')).toBeInstanceOf(Object);
  //   expect(wrapper.find(ScrollingText).prop('film').title);
  // });

});