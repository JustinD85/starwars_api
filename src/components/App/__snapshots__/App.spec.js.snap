// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders without crashing 1`] = `
<div
  className="App"
>
  <ScrollingText
    film={Object {}}
  />
  <FilterSection
    favoriteCount={0}
    setCurrentFilter={[Function]}
  />
  <CardArea
    changeNumber={[Function]}
    currentFilter={null}
    favorites={Array []}
    getData={[Function]}
    max={10}
    min={0}
    toggleFavorite={[Function]}
  />
</div>
`;
